coding/reader.hpp

```diff
@@ -117,7 +82,8 @@ class ReaderPtr
   shared_ptr<TReader> m_p;
 
 public:
-  ReaderPtr(TReader * p = 0) : m_p(p) {}
+  template <typename TReaderDerived>
+  ReaderPtr(unique_ptr<TReaderDerived> && p) : m_p(move(p)) {}
```

Comment:
I know that unique_ptr is move-only. Is it the reason you are asking me not to use &&?